napari/utils/colormaps/colormap.py

```diff
@@ -294,47 +290,75 @@ def _cast_labels_to_minimum_dtype_auto(
     np.ndarray
         Casted labels data.
     """
+    data = _convert_small_ints_to_unsigned(data)
+
+    if data.itemsize <= 2:
+        return data
+
+    num_colors = len(colormap.colors) - 1
+
     dtype = minimum_dtype_for_labels(num_colors + 1)
 
-    return _modulo_plus_one(data, num_colors, dtype, background_value)
+    if colormap.use_selection:
+        return (data == colormap.selection).astype(dtype) * (
+            ((colormap.selection - 1) % num_colors) + 1
+        )
```

Comment:
As above — why not call the `_zero_preserving_modulo` on `np.where(data == colormap.selection, selection, background)`? To me, we really need to avoid repeating code here, as it is the source of many bugs — for each data type, there should be exactly one pipeline from value to color, and the only issue with duplication is whether it occurs in the CPU or the GPU.